src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/DPBasedJoinPlanOptimizer.java

```diff
@@ -62,11 +63,25 @@ public PhysicalPlan getResultingPlan() throws PhysicalOptimizationException {
                         }
 
                         candidatePlans.add( PhysicalPlanFactory.createPlanWithJoin( plan_left,  plan_right) );
-                        if ( plan_left.getRootOperator() instanceof PhysicalOpRequest ) {
-                            PhysicalPlanFactory.enumeratePlansWithUnaryOpFromReq( (PhysicalOpRequest<?,?>) plan_left.getRootOperator(), plan_right, candidatePlans );
+                        final PhysicalOperator rightRootOp = plan_right.getRootOperator();
+                        if ( rightRootOp instanceof PhysicalOpRequest ) {
+                            candidatePlans.addAll( PhysicalPlanFactory.enumeratePlansWithUnaryOpFromReq((PhysicalOpRequest<?, ?>) rightRootOp, plan_left ) );
                         }
-                        else if ( plan_left.getRootOperator() instanceof PhysicalOpRequestWithTranslation ) {
-                            PhysicalPlanFactory.enumeratePlansWithUnaryOpFromReq( (PhysicalOpRequestWithTranslation<?,?>) plan_left.getRootOperator(), plan_right, candidatePlans );
+                        if ( rightRootOp instanceof PhysicalOpRequestWithTranslation ) {
+                            candidatePlans.addAll( PhysicalPlanFactory.enumeratePlansWithUnaryOpFromReq( (PhysicalOpRequestWithTranslation<?,?>) rightRootOp, plan_left ) );
+                        }
+                        if ( rightRootOp instanceof PhysicalOpBinaryUnion || rightRootOp instanceof PhysicalOpMultiwayUnion ){
+                            boolean applicable = true;
+                            for ( int i = 0; i < plan_right.numberOfSubPlans(); i++ ) {
+                                final PhysicalOperator subLop = plan_right.getSubPlan(i).getRootOperator();
+                                if ( !(subLop instanceof PhysicalOpRequest || subLop instanceof PhysicalOpFilter) ) {
```

Comment:
That's not enough. In case that `subLop` is a filter operator, you additionally need to check that there is a request operator directly under that filter.

By the way, why is the program variable called ` subLop`?  I would find `subRootOp` or something like that more natural.